service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/RequestStatementMapper.java

```diff
@@ -41,182 +42,238 @@
     private static final String NOTE = "Annotation: %s @ %s %s";
     private static final String COMMA = ", ";
 
     private final MessageContext messageContext;
     private final CodeableConceptCdMapper codeableConceptCdMapper;
+    private final ParticipantMapper participantMapper;
 
     public String mapReferralRequestToRequestStatement(ReferralRequest referralRequest, boolean isNested) {
-        var requestStatementTemplateParameters = RequestStatementTemplateParameters.builder()
-            .requestStatementId(messageContext.getIdMapper().getOrNew(ResourceType.ReferralRequest, referralRequest.getId()))
-            .isNested(isNested)
-            .availabilityTime(StatementTimeMappingUtils.prepareAvailabilityTimeForReferralRequest(referralRequest))
-            .description(buildDescription(referralRequest))
-            .code(buildCode(referralRequest))
-            .build();
-
-        if (!referralRequest.hasReasonCode()) {
-            requestStatementTemplateParameters.setDefaultReasonCode(DEFAULT_REASON_CODE_XML);
+        return new InnerMapper(referralRequest, isNested).map();
+    }
+
+    @RequiredArgsConstructor
+    private class InnerMapper {
+        private final ReferralRequest referralRequest;
+        private final boolean isNested;
+        private final RequestStatementTemplateParametersBuilder templateParameters = RequestStatementTemplateParameters.builder();
+        private String text = StringUtils.EMPTY;
+
+        private String map() {
+            final IdMapper idMapper = messageContext.getIdMapper();
+            templateParameters
+                .requestStatementId(idMapper.getOrNew(ResourceType.ReferralRequest, referralRequest.getId()))
+                .isNested(isNested)
+                .availabilityTime(StatementTimeMappingUtils.prepareAvailabilityTimeForReferralRequest(referralRequest))
+                .code(buildCode());
+
+            text = buildTextDescription();
+            prependDescriptionInfo();
+
+            if (!referralRequest.hasReasonCode()) {
+                templateParameters.defaultReasonCode(DEFAULT_REASON_CODE_XML);
+            }
+
+            if (referralRequest.hasRequester()
+                    && referralRequest.getRequester().hasAgent()
+                    && referralRequest.getRequester().getAgent().hasReference()) {
+                var requester = referralRequest.getRequester();
+                Reference agentRef = requester.getAgent();
+                var onBehalfOf = requester.hasOnBehalfOf() ? requester.getOnBehalfOf() : null;
+                processAgent(agentRef, onBehalfOf);
+            }
+
+            if (referralRequest.hasRecipient()) {
+                final Reference recipient = referralRequest.getRecipientFirstRep();
+                final var responsibleParty = idMapper.get(recipient);
+                templateParameters.responsibleParty(responsibleParty);
+            }
+
+            templateParameters.text(text);
```

Comment:
For me it is not easy to see where all parts of the text are prepared, the methods have "unexpected side effect"  - building the text is hidden inside different methods.  I would try to build text without state, prepare one method that will take a resource, apply logic and return string value that could be placed in the template. I would stay with the way it was implemented in buildDescription and extend with additional new logic.

Reply:
Agreed. This has been refactored - still pretty involved, but hopefully clearer than before.

Reply:
I prefer the upgraded option now, thank you for changing it.